OpenRA.Game/Map/Map.cs

```diff
@@ -428,6 +430,85 @@ void PostInit()
 				rightDelta + new WVec(0, 0, 512 * ramp[2]),
 				bottomDelta + new WVec(0, 0, 512 * ramp[3])
 			}).ToArray();
+
+			if (MaximumTerrainHeight != 0)
+			{
+				cellProjection = new CellLayer<PPos[]>(this);
+				inverseCellProjection = new CellLayer<List<MPos>>(this);
+
+				// Initialize collections
+				foreach (var cell in AllCells)
+				{
+					var uv = cell.ToMPos(TileShape);
+					cellProjection[uv] = new PPos[0];
+					inverseCellProjection[uv] = new List<MPos>();
+				}
+
+				// Initialize projections
+				foreach (var cell in AllCells)
+					UpdateProjection(cell);
+			}
+		}
+
+		void UpdateProjection(CPos cell)
+		{
+			if (MaximumTerrainHeight == 0)
+				return;
+
+			var uv = cell.ToMPos(TileShape);
+
+			// Remove old reverse projection
+			foreach (var puv in cellProjection[uv])
+				inverseCellProjection[(MPos)puv].Remove(uv);
+
+			var projected = ProjectToShroudPlane(uv).ToArray();
+			cellProjection[uv] = projected;
+
+			foreach (var puv in projected)
+				inverseCellProjection[(MPos)puv].Add(uv);
+		}
+
+		public IEnumerable<PPos> ProjectToShroudPlane(MPos uv)
+		{
+			var mapHeight = MapHeight.Value;
+			if (!mapHeight.Contains(uv))
+				yield break;
```

Comment:
Avoid the compiler generated enumerable and return Enumerable.Empty<PPos>() on this fast path.